Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/IndexingKeyworder.java

```diff
@@ -0,0 +1,502 @@
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kitodo.config.KitodoConfig;
+import org.kitodo.data.database.enums.TaskStatus;
+
+/**
+ * Prepares the search keywords for a process or task.
+ */
+class IndexingKeyworder {
+    private static final Logger logger = LogManager.getLogger(IndexingKeyworder.class);
+
+    private static final String PSEUDOWORD_TASK_AUTOMATIC = "automatic";
+    private static final String PSEUDOWORD_TASK_DONE = "closed";
+    private static final String PSEUDOWORD_TASK_DONE_PROCESSING_USER = "closeduser";
+    private static final String ANY_METADATA_MARKER = "mdWrap";
+    private static final char VALUE_SEPARATOR = 'q';
+
+    private static final Pattern TITLE_GROUPS_PATTERN = Pattern.compile("[\\p{IsLetter}\\p{Digit}]+");
+    private static final Pattern METADATA_PATTERN = Pattern.compile("name=\"([^\"]+)\">([^<]*)<", Pattern.DOTALL);
+    private static final Pattern METADATA_SECTIONS_PATTERN =
+            Pattern.compile("<mets:dmdSec.*?o>(.*?)</kitodo:k", Pattern.DOTALL);
+    private static final Pattern RULESET_KEY_PATTERN =
+            Pattern.compile("key id=\"([^\"]+)\">(.*?)</key>", Pattern.DOTALL);
+    private static final Pattern RULESET_LABEL_PATTERN = Pattern.compile("<label[^>]*>([^<]+)", Pattern.DOTALL);
+
+    private static final Map<String, Map<String, Collection<String>>> rulesetCache = new HashMap<>();
+
+    private Set<String> titleKeywords = Collections.emptySet();
+    private Set<String> projectKeywords = Collections.emptySet();
+    private Set<String> batchKeywords = Collections.emptySet();
+    private Set<String> taskKeywords = Collections.emptySet();
+    private Set<String> taskPseudoKeywords = Collections.emptySet();
+    private Set<String> metadataKeywords = Collections.emptySet();
+    private Set<String> metadataPseudoKeywords = Collections.emptySet();
+    private String processId = null;
+    private Set<String> commentKeywords = Collections.emptySet();
+
+    public IndexingKeyworder(Process process) {
+        this.titleKeywords = filterMinLength(initTitleKeywords(process.getTitle()));
+        String projectTitle = Objects.nonNull(process.getProject()) ? process.getProject().getTitle() : "";
+        this.projectKeywords = filterMinLength(initSimpleKeywords(projectTitle));
+        this.batchKeywords = filterMinLength(initBatchKeywords(process.getBatches()));
+        var taskKeywords = initTaskKeywords(process.getTasksUnmodified());
+        this.taskKeywords = filterMinLength(taskKeywords.getLeft());
+        this.taskPseudoKeywords = filterMinLength(taskKeywords.getRight());
+        String place = null;
+        if (logger.isDebugEnabled()) {
```

Comment:
Why is this following only executed if log level is on debug or lower?

Reply:
The string is only created to be used in logging when DEBUG logging is enabled. So, in production, it would not need to be created since it is not used.

Reply:
Thank you for the explanation. Maybe this should be added as a comment as I think that would cause questions in future.